### PR TITLE
Add tests for FAISS and release memory

### PR DESCRIPTION
**Pull request by koparasy**

### CODE REVIEW

File: examples/main.cpp [hunk 1]
```diff
@@ -700,6 +700,14 @@ int main(int argc, char **argv)
     CALIPER(CALI_MARK_END("Cycle");)
     MPI_CALL(MPI_Barrier(MPI_COMM_WORLD));
   }
+
+  delete [] workflow;
+
+  for ( int mat_idx = 0; mat_idx < num_mats; ++mat_idx){
+    delete eoses[mat_idx];
```
Comment (ggeorgakoudis) [inline examples/main.cpp:708]: I'll accept this but add a TODO to use smart pointers

File: src/ml/hdcache.hpp [hunk 2]
```diff
@@ -8,6 +8,8 @@
 #ifndef __AMS_HDCACHE_HPP__
 #define __AMS_HDCACHE_HPP__
 
+#include <bits/stdc++.h>
```
Comment (ggeorgakoudis) [inline src/ml/hdcache.hpp]: This is GNU specific. Is that desirable?

File: src/ml/hdcache.hpp [hunk 3]
```diff
@@ -248,7 +250,11 @@ class HDCache
     return new_cache;
   }
 
-  ~HDCache() { DBG(Surrogate, "Destroying UQ-cache") }
+  ~HDCache() {
+    DBG(Surrogate, "Destroying UQ-cache");
```
Comment (ggeorgakoudis) [inline src/ml/hdcache.hpp]: I'd rename that since UQ can be performance even without an HDCache

File: src/ml/hdcache.hpp [hunk 4]
```diff
@@ -546,12 +558,15 @@ class HDCache
         if (m_policy == AMSUQPolicy::FAISSMean) {
           total_dist =
               std::accumulate(kdists + i * knbrs, kdists + (i + 1) * knbrs, 0.);
-          is_acceptable[i] = (ook * total_dist) < acceptable_error;
+          total_dist = ook * total_dist;
```
Comment (ggeorgakoudis) [inline src/ml/hdcache.hpp]: `total_dist` is a misnomer in this case, more descriptive is `mean_dist`

File: src/ml/hdcache.hpp [hunk 5]
```diff
@@ -546,12 +558,15 @@ class HDCache
...
+          is_acceptable[i] = total_dist < acceptable_error;
         } else if (m_policy == AMSUQPolicy::FAISSMax) {
           // Take the furtherst cluster as the distance metric
-          total_dist = kdists[i * knbrs + knbrs - 1];
+          total_dist = *std::max_element(&kdists[i * knbrs],
```
Comment (ggeorgakoudis) [inline src/ml/hdcache.hpp]: `total_dist` is a misnomer, more descriptive is `max_dist`

File: src/wf/cuda/utilities.cuh
Comment (ggeorgakoudis) [inline src/wf/cuda/utilities.cuh:1]: Is this only formatting changes?

File: tests/generate_faiss.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    # xq_true = create_elements(args.num_centers,args.dimensions)
+    # xq = np.vstack((xq_false, xq_true)).astype('float32')
+    # print("Search for")
+    # print(xq)
+    # D, I = index.search(xq, k)
+    # print("Distance is", D)
+    # print("Indexes are", I)
+
+    # predicate = np.any(D < 4, axis=1)
+    # print(predicate)
```
Comment (ggeorgakoudis) [inline tests/generate_faiss.py:69]: Are comments needed? Are they performing the same function as C++? Maybe useful to have a mode that runs the test in pure python through an input flag.

File: tests/test_hdcache.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+  const T offset = 5.0;
+  for (int i = 0; i < dims; i++) {
+    T *data = ams::ResourceManager::allocate<T>(num_clusters * elements,
+                                                AMSResourceType::HOST);
+    for (int j = 0; j < elements; j++) {
+      // Generate a value for every cluster center
+      for (int k = 0; k < num_clusters; k++) {
+        T tmp = ((T)rand()) / INT_MAX;
+        tmp += (k + 1) * num_clusters;
+        if (j % 2 == 0) {
```
Comment (ggeorgakoudis) [inline tests/test_hdcache.cpp]: Clearer as `if ((j % 2) == 0)`

File: tests/test_hdcache.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+template <typename T>
+void print_vectors(std::vector<T *> &vec, int num_elements, int num_clusters)
+{
+  for (int i = 0; i < num_elements; i++) {
+    for (int c = 0; c < num_clusters; c++) {
+      for (auto v : vec) {
+        std::cout << v[i * num_clusters + c] << ":";
+      }
+      std::cout << std::endl;
```
Comment (ggeorgakoudis) [inline tests/test_hdcache.cpp]: Is there a particular reason to flush the output here with `std::endl`?

File: tests/test_hdcache.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+  bool res = validate(nClusters, nElements, h_predicates);
+
+  ams::ResourceManager::deallocate(h_predicates, AMSResourceType::HOST);
+  return res;
+}
+
+
+int main(int argc, char *argv[])
+{
+  using namespace ams;
```
Comment (ggeorgakoudis) [inline tests/test_hdcache.cpp:136]: Abort if there are missing arguments and print usage.

**Review by ggeorgakoudis — CHANGES_REQUESTED** [review]

File: tests/generate_faiss.py (hunk 6 above)
Comment (koparasy) [reply tests/generate_faiss.py:69]: That is a good idea, but currently I don't have a way to generate an index for arbitrary dimensions and being certain I can write a passing test for that arbitary selected dimension. I will need to think about this more.

File: src/wf/cuda/utilities.cuh
Comment (koparasy) [reply src/wf/cuda/utilities.cuh:1]: Yes

File: examples/main.cpp (hunk 1 above)
Comment (koparasy) [reply examples/main.cpp:708]: Fixed

File: src/ml/hdcache.hpp (hunk 2 above)
Comment (koparasy) [reply src/ml/hdcache.hpp]: Removed

File: tests/test_hdcache.cpp (hunk 8 above)
Comment (koparasy) [reply tests/test_hdcache.cpp]: Fixed

File: tests/test_hdcache.cpp (hunk 9 above)
Comment (koparasy) [reply tests/test_hdcache.cpp:136]: Done

File: tests/test_hdcache.cpp (hunk 7 above)
Comment (koparasy) [reply tests/test_hdcache.cpp]: Fixed

File: tests/generate_faiss.py (hunk 6 above)
Comment (ggeorgakoudis) [reply tests/generate_faiss.py:69]: Agreed

**Review by ggeorgakoudis — APPROVED** [review]

**Pull request merged**